File: sweetfactory/src/main/java/sweetfactory/SweetFactory.java
package sweetfactory;

import org.jetbrains.annotations.Nullable;

public class SweetFactory {

    public static @Nullable <T> T newInstanceOf(Class<T> clazz) {
        final String className = clazz.getName() + "Impl";
        try {
            final Class c = Class.forName(className);
            return (T) c.newInstance();
        } catch (Exception e) {
            return null;
        }
    }
}

File: sweetfactory/src/main/java/sweetfactory/SweetFactory.kt
package sweetfactory

import kotlin.reflect.KClass

/**
 * Create an instance of a factory implemented by SweetFactory processor
 * <pre>`
 * SweetFactory.newInstanceOf(FeatureActivityFactory::class)
`</pre> *
 */
object SweetFactory {

    @JvmStatic
    fun <T : Any> newInstanceOf(clazz: KClass<T>): T? = newInstanceOf(clazz.java)

    @JvmStatic
    @Suppress("UNCHECKED_CAST")
    fun <T> newInstanceOf(clazz: Class<T>): T? {
        return try {
            val className = clazz.name + "Impl"
            val c = Class.forName(className)
            c.newInstance() as T
        } catch (ex: ClassNotFoundException) {
            null
        }
    }

}
